cio/src/configs.rs

```diff
@@ -72,6 +73,84 @@ pub struct Config {
     pub certificates: BTreeMap<String, NewCertificate>,
 }
 
+#[derive(Debug, Deserialize, Clone, JsonSchema, Serialize, PartialEq, FromSqlRow, AsExpression)]
+#[serde(rename_all = "lowercase")]
+#[diesel(sql_type = VarChar)]
+pub enum ExternalServices {
+    Airtable,
+    GitHub,
+    Google,
+    Okta,
+    Ramp,
+    Zoom,
+}
+
+impl ExternalServices {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            ExternalServices::Airtable => "airtable",
+            ExternalServices::GitHub => "github",
+            ExternalServices::Google => "google",
+            ExternalServices::Okta => "okta",
+            ExternalServices::Ramp => "ramp",
+            ExternalServices::Zoom => "zoom",
+        }
+    }
+
+    pub async fn get_provider_writer(
+        &self,
+        db: &Database,
+        company: &Company,
+    ) -> Result<Box<dyn ProviderWriteOps + Send + Sync>> {
+        Ok(match self {
+            // We don't need a base id here since we are only using the enterprise api features.
+            ExternalServices::Airtable => Box::new(company.authenticate_airtable("")),
+            ExternalServices::GitHub => Box::new(company.authenticate_github()?),
+            ExternalServices::Google => Box::new(company.authenticate_google_admin(db).await?),
+            ExternalServices::Okta => Box::new(
+                company
+                    .authenticate_okta()
+                    .ok_or_else(|| anyhow::anyhow!("Failed to instantiate Okta client"))?,
+            ),
+            ExternalServices::Ramp => Box::new(company.authenticate_ramp(db).await?),
+            ExternalServices::Zoom => Box::new(company.authenticate_zoom(db).await?),
+        })
+    }
+}
+
+impl fmt::Display for ExternalServices {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ExternalServices::Airtable => write!(f, "Airtable"),
+            ExternalServices::GitHub => write!(f, "GitHub"),
+            ExternalServices::Google => write!(f, "Google"),
+            ExternalServices::Okta => write!(f, "Okta"),
+            ExternalServices::Ramp => write!(f, "Ramp"),
+            ExternalServices::Zoom => write!(f, "Zoom"),
+        }
+    }
+}
+
+impl ToSql<VarChar, Pg> for ExternalServices {
+    fn to_sql<W: std::io::Write>(&self, out: &mut Output<W, Pg>) -> serialize::Result {
+        <str as ToSql<VarChar, Pg>>::to_sql(self.as_str(), out)
+    }
+}
+
+impl FromSql<VarChar, Pg> for ExternalServices {
+    fn from_sql(bytes: PgValue<'_>) -> deserialize::Result<Self> {
+        match bytes.as_bytes() {
+            b"airtable" => Ok(ExternalServices::Airtable),
+            b"github" => Ok(ExternalServices::GitHub),
+            b"google" => Ok(ExternalServices::Google),
+            b"okta" => Ok(ExternalServices::Okta),
+            b"ramp" => Ok(ExternalServices::Ramp),
+            b"zoom" => Ok(ExternalServices::Zoom),
+            _ => Err("Unrecognized enum variant".into()),
```

Comment:
[non-blocking] It might be a good idea to make this error message to be a bit clearer as to what actually happened. You'll be able to look it up in the stack trace, but at least logging (if possible) the name of the service that it tried to render might be useful.

Reply:
Agreed, added a better error message denoting that deserialization failed and an attempt to turn the received bytes into a string representation.